File: Assignment1/nitin_sangar_task1.py
from pyspark.sql import SparkSession
from pyspark import SparkContext
import os

import json
import operator
import sys


inputPath = sys.argv[1]
outputPath = sys.argv[2]


sc = SparkContext('local[*]','nitin_sangar_sc')
# Task 1
data = sc.textFile(inputPath).map(json.loads).map(lambda x:(x['useful'], x['stars'], x['text'], x['user_id'], x['business_id']))

result = {}
#Useful reviews
usefulReviews = data.filter(lambda x: x[0] > 0)
#print('\nNumber of Useful Reviews : {}'.format(usefulReviews.count()))

result["n_review_useful"] = usefulReviews.count()

#Number of reviews that have five star rating
fiveRatingReviews = data.filter(lambda x: x[1] == 5).count()
#print('\nNumber of reviews with 5 star : {}'.format(fiveRatingReviews.count()))
result["n_review_5_star"] = fiveRatingReviews

#characters in the text of longest review
maxReview = data.max(lambda x:len(x[2]))
#print("\nCharacters in Max Review : {}".format(len(maxReview['text'])))
result['n_characters'] = len(maxReview[2])

#Number of distinct users who wrote reviews
distinctCount = data.map(lambda x:x[3]).distinct().count()
#print("\nNumber of Distinct Users : {}".format(distinctCount))
result['n_user'] = distinctCount

#Top 20 users who wrote the largest number of Reviews
users = data.map(lambda x:(x[3],1))
userRDD = users.countByKey()
sorted_Users = sorted(userRDD.items(), key=lambda e:(-e[1],e[0]))
i = 0;
userData = []
#print('\n Top 20 Users - ')
for user,reviews in sorted_Users:
    userData.append([user, reviews])
    i += 1
    if i == 20:
        break;
result['top20_user'] = userData

#Number of distinct business that have been reviewed
distinctBusiness = data.map(lambda x: x[4]).distinct().count()
result['n_business'] = distinctBusiness

#Top 20 business
business = data.map(lambda x: (x[4],1))
businessRDD = business.countByKey()
sorted_business = sorted(businessRDD.items(), key=lambda e:(-e[1],e[0]))
i = 0;
businessData = []
#print('\n Top 20 Business - ')
for business,reviews in sorted_business:
    businessData.append([business, reviews])
    i += 1
    if i == 20:
        break;
result['top20_business'] = businessData
with open(outputPath, 'w') as outfile:
    json.dump(result, outfile, indent=2)

File: Assignment1/nitin_sangar_task1.scala
import java.io.FileWriter
import org.apache.spark.sql.{Row, SparkSession}

object nitin_sangar_task1 {
  def main(args: Array[String]): Unit = {

    val spark = SparkSession.builder
      .master("local[*]")
      .appName("Spark Word Count")
      .getOrCreate()

    val inputPath = args(0)
    val outputPath = args(1)


    val sc = spark.sparkContext;

    val data = spark.read.json(inputPath).rdd

    // Useful reviews
    val usefulReviews = data.filter(x => (x.getAs[Long]("useful") > 0)).count()

    val fiveStarsCount = data.filter(x => x.getAs[Double]("stars") == 5).count()

    val maxReview = data.map(x => x.getAs[String]("text").length()).max();

    val distinctUsers = data.map(x => x.getAs[String]("user_id")).distinct().count();

    val users = data.map(x => (x.getAs[String]("user_id"), 1))
    val userRdd = users.countByKey()
    val sortedUsers = userRdd.toSeq.sortWith { (x, y) => (x._2 > y._2) || ((x._2 == y._2) && x._1 < y._1) }.take(20).toList

    val distinctBusiness = data.map(x => x.getAs[String]("business_id")).distinct().count();

    val business = data.map(x => (x.getAs[String]("business_id"), 1))
    val businessRdd = business.countByKey()
    val sortedBusiness = businessRdd.toSeq.sortWith { (x, y) => (x._2 > y._2) || ((x._2 == y._2) && x._1 < y._1) }.take(20).toList

    val file = new FileWriter(outputPath)
    file.write("{")
    file.write("\n\"n_review_useful\": " + usefulReviews + ",")
    file.write("\n\"n_review_5_star\": " + fiveStarsCount + ",")
    file.write("\n\"n_characters\": " + maxReview + ",")
    file.write("\n\"n_user\": " + distinctUsers + ",")
    file.write("\n\"top20_user\": [");
    var i = 1
    for ((k, v) <- sortedUsers) {
      file.write("[")
      file.write("\"" + k.toString() + "\"")
      file.write("," + v)
      i += 1
      if (i == 21) {
        file.write("]")
      } else {
        file.write("],")
      }
    }
    file.write("],")
    file.write("\n\"n_business\": " + distinctBusiness + ",")
    file.write("\n\"top20_business\": [");
    i = 1
    for ((k, v) <- sortedBusiness) {
      file.write("[")
      file.write("\"" + k.toString() + "\"")
      file.write("," + v)
      i += 1
      if (i == 21) {
        file.write("]")
      } else {
        file.write("],")
      }
    }
    file.write("]")
    file.write("\n}")
    file.flush()
  }
}
